### PR TITLE
Crypto-15: Create DetailsScreen

### PR DESCRIPTION
**Pull request by vitalyaliskevych**
### Whats News
- Create UI from DetailScreen(without graph component)
### Issues
- [Crypto-15](https://trello.com/c/1w82ctqy)
### Screenshot
Screenshot 2023-01-31 at 15 41 39

### CODE REVIEW

File: Crypto/Extentions/Toolbar.swift [hunk 1]
```diff
@@ -18,4 +18,24 @@ extension View {
                 }
             })
         }
-}
+    func createToolBarDetailView(image: String, text: String, dismissAction: (() -> Void)?) -> some View {
+            self.toolbar(content: {
```
Comment (vasyl-nadtochii) [inline Crypto/Extentions/Toolbar.swift:22]: bad alignment

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+//  DetailScreenView.swift
+//  Crypto
+//
+//  Created by Developer on 06.12.2022.
+//
+
+import SwiftUI
+
+struct DetailScreenView: View {
+    @ObservedObject var viewModel = DetailViewModel()
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:11]: add 1 space below and abowe

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    var body: some View {
+        NavigationView {
+            ZStack {
+                Color.mainColor.opacity(0.9)
+                    .ignoresSafeArea()
+                VStack {
+                    ScrollView {
+                        main
+                    }
+                    createButton(text: "Buy Bitcoin", dismissAction: {})
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:21]: text should be localized
it should accept Coin's name as a parameter
`"Buy [coin]"`

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+private extension DetailScreenView {
+    var main: some View {
+        VStack(spacing: 15) {
+            bitcoinPrice
+            news
+        }
+    }
+    var bitcoinPrice: some View {
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:40]: 1. add space above
2. It shouldn't be named `bitcoinPrice`. This screen can accept any coin from list, not only Bitcoin. So it's better to name it `currentPrice`

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            .background(Color.black.opacity(0.2))
+            .cornerRadius(15)
+            Spacer()
+        }
+        .padding()
+    }
+    var news: some View {
+        VStack {
+            HStack {
+                Text("News")
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:58]: should be localized

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                    .foregroundColor(Color.white)
+                    .font(.system( size: 26, weight: .bold))
+                    .padding()
+                Spacer()
+            }
+            ForEach(viewModel.newsModels, id: \.id) { newsModel in
+                createToNews(newsModel: newsModel)
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:65]: `createNewsRow` is better I guess

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                Spacer()
+            }
+            HStack {
+                Text(newsModel.newsDate)
+                    .font(.system( size: 18, weight: .medium))
+                    .foregroundColor(Color.gray.opacity(0.8))
+                    .padding(.leading, 20)
+                Spacer()
+            }
+            Divider()
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:86]: for the last news row we need divider to be hidden. Please, handle that

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        }
+        .padding(.top, 20)
+        
+    }
+    
+    func createButton(text: String, dismissAction: (() -> Void)?) -> some View {
+        VStack {
+            Button(action: {
+                dismissAction?()
+                }) {
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:98]: bad alignment

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            }
+            Divider()
+                .frame( width: UIScreen.main.bounds.width / 1.1, height: 1)
+                .background(Color.gray)
+        }
+        .padding(.top, 20)
+        
+    }
+    
+    func createButton(text: String, dismissAction: (() -> Void)?) -> some View {
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:94]: 1. Better to name `createBuyCoinButton`
2. `dismissAction` ->`action`

**Review by vasyl-nadtochii — CHANGES_REQUESTED** [review]
1. Resolve comments
2. The main thing about this screen is that it can accept `CoinInfo` object as a parameter. This object should contain prices (for 5 days), news etc. Please handle that in appropriate way

File: Crypto/Extentions/Toolbar.swift [hunk 10]
```diff
@@ -30,6 +50,25 @@ extension View {
...
             }
+            ToolbarItem(placement: .principal) {
+                Text(text)
+                    .font(.system(size: 35, weight: .bold))
+                    .foregroundColor(.white)
+            }
+        })
+    }
+    
+    func createToolBarDetailView(image: String, text: String, dismissAction: (() -> Void)?) -> some View {
```
Comment (vasylnadtochii) [inline Crypto/Extentions/Toolbar.swift:61]: do we need that func?

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    var body: some View {
+        NavigationView {
+            ZStack {
+                Color.mainColor.opacity(0.9)
+                    .ignoresSafeArea()
+                VStack {
+                    ScrollView {
+                        main
+                    }
+                    createBuyCoinButton(text: "buy" + String(" " + coinName.name), action: {})
```
Comment (vasylnadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:24]: `text` should be created using `localizedWithVars` func

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    var main: some View {
+        VStack(spacing: 15) {
+            coinPrice
+            news
+        }
+    }
+    var coinPrice: some View {
+        HStack {
+            HStack {
+                Text("$" + String(format:"%.2f", coinName.price))
```
Comment (vasylnadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:45]: create an extension for `Double` which will return formatted price

File: Crypto/Resources/en.lproj/Localizable.strings [hunk 13]
```diff
@@ -10,3 +10,5 @@
 "our_website" = "Our website";
 "notification" = "Notification";
 "version" = "Version";
+"news" = "News";
+"buy" = "Buy, %@";
```
Comment (vasylnadtochii) [inline Crypto/Resources/en.lproj/Localizable.strings:14]: `,` is redundant

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+//  Created by Developer on 06.12.2022.
+//
+
+import SwiftUI
+
+struct DetailScreenView: View {
+    
+    @ObservedObject var viewModel: DetailViewModel
+    
+    var coinName: Coin
```
Comment (vasylnadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:14]: it should be in `DetailScreenViewModel`

**Review by vasylnadtochii — CHANGES_REQUESTED** [review]
1. Resolve comments
2. Check how `localizedWithVars` works
3. You should have in your view structs only one field - `ViewModel` (it's in general, but some exceptions are also possible). Do not store any other data in your views

File: Crypto/Extentions/Toolbar.swift (hunk 10 above)
Comment (vitalyaliskevych) [reply Crypto/Extentions/Toolbar.swift:61]: yes, we have a toolbar on the screen

Comment (vitalyaliskevych) [reply Crypto/Extentions/Toolbar.swift:61]: no, i remoov this

File: Crypto/App/CryptoApp.swift [hunk 15]
```diff
@@ -11,7 +11,6 @@ import SwiftUI
 struct CryptoApp: App {
     var body: some Scene {
         WindowGroup {
-           SettingsView(viewModel: SettingsViewModel())
-        }
+            IntroView()        }
```
Comment (vasyl-nadtochii) [inline Crypto/App/CryptoApp.swift:14]: } move to next line

File: Crypto/Extentions/Double.swift [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+//  Crypto
+//
+//  Created by Developer on 09.02.2023.
+//
+
+import Foundation
+
+extension Double {
+    func toString() -> String {
+        return String(format: "%.2f",self)
```
Comment (vasyl-nadtochii) [inline Crypto/Extentions/Double.swift:12]: add space after comma

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+            }
+            createDivider
+        }
+        .padding(.top, 20)
+    }
+    
+    func createBuyCoinButton(text: String, action: (() -> Void)?) -> some View {
+        VStack {
+            Button(action: {action?()}) {
+                HStack {
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:97]: do we need this HStack?

File: Crypto/Flows/Screens/DetailScreen/DetailView.swift [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+
+struct DetailScreenView_Previews: PreviewProvider {
+    static var previews: some View {
+        DetailScreenView(viewModel: DetailViewModel(coinName: .init(name: "Bitcoin", price: 1651.64, icon: Image.btcIcn)))
+    }
+}
+
+
+
+
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailView.swift:125]: remove redundant spaces

File: Crypto/Flows/Screens/DetailScreen/DetailViewModel.swift [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+    ]
+    let coinModels: [Coin] = [
+        .init(name: "Bitcoin", price: 21188.04, icon: .btcIcn),
+        .init(name: "Litecoin", price: 68.06, icon: .ltcIcn),
+        .init(name: "Ethereum", price: 1651.64, icon: .ethIcn),
+        .init(name: "Bitcoin", price: 21188, icon: .btcIcn),
+        .init(name: "Litecoin", price: 68.06, icon: .ltcIcn),
+        .init(name: "Ethereum", price: 1651, icon: .ethIcn),
+    ]
+
```
Comment (vasyl-nadtochii) [inline Crypto/Flows/Screens/DetailScreen/DetailViewModel.swift:28]: redundant space

**Review by vasyl-nadtochii — APPROVED** [review]
Please fix last comments and then merge

**Pull request merged**